### ISSUE TITLE
[FEATURE REQ] SQL Load recent history only (i.e., for real-time)

### ISSUE THREAD

**Issue by markelwin** [browser-history/browser-history#284]
I am cc'ing the maintainers of this project (@Samyak2 and @ObaraEmmanuel since 2020).

**Is your feature request related to a problem? Please describe.**
I wish to use your package to provide real-time inspection into recent (last hour) browser history for the purpose of the development of a new application categorization. For instance, as you know, combination of `browser-history` with a cookie reading package, such as `browsercookie`, allows any Python application to access the recent **specific** pages viewed by any user, and simple three-line commands from `AppKit` allow any application to detect when the user pushes focus into, and out of, a browser. For instance, as I software engineer using your package, I may push focus over to a browser, search for answers, and return to my engineering development error, and a combination of these three resources allows my software to identify the specific sites visited, specific queries pushed into search engines, and specific information which my eyes used: machine reading. 

**Describe the solution you'd like**
The package I am building is described above and utilized additional features such as my software package [here](https://github.com/markelwin/CloudNode) which is unrelated to this request. My firm wishes to publish to the open source a simple applet which identifies recent browser information during normative OS application switching. Reading and writing in a word processor is another use case. 

**Describe alternatives you've considered**
We have all the pieces functional. However, the current implementation of `browser-history`, which we use, only allows for the `Browser().fetch_history(desc=True)` use case in which all history is loaded, instead of a narrow window (which could be injected directly into the SQL inquiry the `.fetch_history()` routine makes. This would purposefully increase the load time from minutes to near seconds, allowing for real-time operation and connection with the other packages described. Presumably some measure of caching within the `browser-history` itself could be implemented for repeated calls; however, the end user caching the last known window of operational load would be sufficient to perform the same task as long as `.fetch_history()` accepted an SQL window for its retrieval function call immediately.

**Additional context**
Pretty remarkable that only one package for this utility exists, and we should be grateful for having even that, in this competitive tech environment. Please consider reaching out directly to me to discuss, quite frankly, whatever else you would like. Thank you. Please note that we intend to use Mozilla Firefox and other open-source tools yet be cross-platform.

<img width="942" alt="Screenshot 2024-09-26 at 8 42 27 PM" src="https://github.com/user-attachments/assets/0d056cd1-52ce-4fa8-acb1-5991f9d335f7">

**Comment by Samyak2** [comment]
Seems like a very useful feature. Thank you for the suggestion! I'll check and see what it will take to build this

**Comment by markelwin** [comment]
<img width="759" alt="Screenshot 2024-09-29 at 12 45 45 PM" src="https://github.com/user-attachments/assets/177de987-6243-45b0-8851-450a234b1b62">

I agree @Samyak2. It appears to me that the `sqlite3` connection makes a single `.fetchall()` call for the straightforward, hard-coded `history_SQL` SQL query, described above. It appears to me that the SQL only needs to include a select filter on `visit_time` for the window of time preferred (i.e., the most recent hour). 

I subsequently also notice that the browser history SQL database file is copied to a temporarily directory. Is this necessary because of file access permissions (i.e., in /Application Support/ on MacOS)? These could easily be cached on the search object (the variable `f` in my original post) unless you want to get fancy and implement a session context manager, which does not seem necessary to deliver the feature discussed here. 


PS: I could perhaps attempt the pull request myself with your permission. I need the feature only in about one week time.
PS: If either of you want to help me build this larger project I would certainly delight in the collegialness.

**Comment by Samyak2** [comment]
@markelwin it would be great and very welcome if you're willing to work on this feature! Unfortunately, I'm a little low on time these days to attempt it myself. Thank you in advance for the contribution.

> I subsequently also notice that the browser history SQL database file is copied to a temporarily directory. Is this necessary because of file access permissions (i.e., in /Application Support/ on MacOS)? These could easily be cached on the search object (the variable `f` in my original post)

Another reason to do that is that, on windows the file cannot be read while it's open in another app. The temp file could be cached in the `Browser` class' constructor, but that would be a change in behavior and needs to be documented. If the browser object is long-lived, the results will be stale since the copy would only be made at the time of initialization. Essentially, it introduces a problem of cache invalidation. I don't mind this solution as long as it's documented and there's a param to force a copy. @ObaraEmmanuel thoughts?

**Comment by ObaraEmmanuel** [comment]
Thanks for your interest @markelwin. I think it would be better to have the caching implemented by the respective application. We should try to keep the package simple IMO. We also wouldn't want to break backward compatibility like this even if it's documented. 
CC: @Samyak2

**Comment by Samyak2** [comment]
Good points @ObaraEmmanuel.

@markelwin the overhead of the copy should be small. Are you seeing any significant slowdown because of it? If not, I think we should focus on the recent history feature first.

**Comment by markelwin** [comment]
@ObaraEmmanuel @Samyak2 The three of us are in agreement. I should have some time within the next few days.

**Comment by markelwin** [comment]
Quite good news. I was able to make three non-destructive changes to the mechanism by which the package transfers history files to a temporarily directory and to the SQL query itself (which in total account for 100% of the runtime) to bring recurrent calls for histories (i.e., 'the last hour of browsing data') down from 3.1s to 0.09s (on my machine), and that do not change the API call for any of the functions. 

I noticed three important details of the code structure to drive these changes.

1. `.fetch_history()` is actually called in several places, which makes any injection of a new parameter (i.e., `window_ts`) a laborious change that would have to propagate across the code; however, the `.history_SQL` query itself is a property to a constant attribute, and only a property so that custom browsers can construct whatever they want. This means that a user can simply populate the `.history_SQL` attribute of the `Browser` object directly with whatever SQL query they want external to the `.fetch_history()` call, and that is what I do for this solution. What I then recommend is that I add to the package a standard SQL helper which generates the two use cases (all history; using `window_ts`). Here is pseudo-code:

```python
f = Firefox()
f.history_SQL = f.configure_query(window_ts=[start_ts, end_ts])
results = f.fetch_history()
```

The changes to the SQL query (e.g. in Chrome) are to change its where statement from 
```
WHERE
    visits.visit_duration > 0
```
to
```
WHERE
    datetime(visits.visit_time/1000000-11644473600, 'unixepoch', 'localtime') BETWEEN '2024-10-06 07:00:00' AND '2024-10-06 09:00:00'
    AND
    visits.visit_duration > 0
```
Performing the datetime WHERE before the visit_duration WHERE is also slightly faster (about 10% compared to the reverse). The SQL query is specific for each Browser object and so I believe the new method should be on the Browser object itself as show above. We could also construct this as a context manager so that the 'all histories' default is resupplied after the context manager ends. (This is much unnecessary in my opinion.) This accounts for an increase of speed from 2.8s to 0.28s on my machine (to retrieve one hour of recent searches) and is roughly proportional to the length of the window (one year of recent searches takes about 1.3s, etc.)

2. The method by which the `.fetch_history()` transfers the History files to temporary directory can be improved within `.fetch_history()` without caching, and is cleaner and more sensible for its existing purpose. The change is two parts: using the standard temporary directory and supplying each history `history_path` file a unique temporary filename based on the hash of its absolute path (so that repeated copies always land on the same temporary file); and not copying the file from system to temporary when the file already exists and has the same modified timestamp (`os.path.getmtime`). What this does is allow repeated calls to only update the most recently changed history file, if any, without relying on a specified cache or user-controlled cache mechanism (the cache system is implicit). This increased speed from 3.1s to 0.09s, very quick. It has only a small effect on the retrieval of all histories (as 0.20s accounts for a small fraction of the 3.1s query time). Here is the code change inside `.fetch_history()`

```python
                # START NEW CODE
                import hashlib
                tmpdirname = tempfile.gettempdir()  # override the choice to use a new temporary directory each time.
                temporary_basename = hashlib.md5(str(history_path.absolute()).encode()).hexdigest()
                copied_history_path = os.path.join(tmpdirname, temporary_basename)
                if os.path.exists(copied_history_path):
                    system_ts = os.path.getmtime(history_path.absolute())
                    copied_ts = os.path.getmtime(copied_history_path)
                    if system_ts != copied_ts:  # if any difference exists default to the system copy
                        shutil.copy2(history_path.absolute(), copied_history_path)
                else: shutil.copy2(history_path.absolute(), copied_history_path)  # copy if temporary does not exist
                # copied_history_path = shutil.copy2(history_path.absolute(), tmpdirname)
                # END NEW CODE
```

In the total implementation the `tmpdirname` would simply be set as such above and not be overwritten, as one might expect.

3. The use of `sort` for the six lines from `if sort:` onward can be moved to the outermost scope immediately before the return. As currently configured the sort method is called each time a new history file is added to the results (which is redundant, and moderately time consuming). I have confirmed that the results are properly sorted with the change. This change is significant for retrieving all histories (as you would imagine) and speed up all history searches from 3.1s to about 2.7s (about 15%). As you can imagine there was nearly no change in speed for the real-time recent search use case.

This all results in no change to the API calls and adds one new method to the Browser class to construct windowed queries.

Speed improvements: 
all history (about three years): 3.1s to 2.0s
last hour: 3.1s to 0.09s

What do you think?

**Comment by Samyak2** [comment]
@markelwin the results look great! Thank you for the effort you put in this. Would you mind opening a PR with these improvements? I have some feedback already though.

1. The `configure_query` API seems a bit awkward. The user shouldn't have to know about history_SQL to use the class. Is there any other disadvantage to adding a new param? Other than it being a change across many functions. If not, I'll be willing to help you make the changes. An API like `f.fetch_history(window=[start_ts, end_ts])` would be much preferred.
2. This approach sounds good. Although, considering that we're depending on `shutil.copy2` preserving the mtime to the copy, I would want to check if this behavior is consistent on all systems (Mac, Linux, Windows).
3. Good catch!

**Comment by markelwin** [comment]
@Samyak2 Yes. Quite happy with the results. I will open a PR later this evening or tomorrow. 

1. Agreed. I strongly suggest an alternative design pattern that places encapsulates the API on the Browser object instead. I believe the API and code is cleaner, and since the logic for the SQL is on the Browser object its access point should remain there, just as any other configuration might, especially since construction of the browser objects is nearly zero, and I would use datetime instead of timestamps to keep consistent with how you return results already.

``` python
f = Firefox()
f.fetch_history()  # all results
f = Firefox().history_range(window=[start_dt, end_dt]) 
results = f.fetch_history()
```

Can we do this?

2. Agreed.
3. Yeah.

**Comment by Samyak2** [comment]
I agree with using datetime instead of timestamp. I'm not sure I follow the rest.

Instead, I have a proposal. Let me know if it is consistent with what you were thinking.

I agree with keeping the window-specific configuration alongside the `history_SQL`.

Firstly, we modify the `history_SQL` to be a `.format()`-able string like (the `...` represents the existing SQL):
```python
history_SQL = """
select ...
from ...
where
    ...
    {window_part}
order by ...
"""
```

And we add a new `window_SQL` property that has something like (note the placeholders instead of actual values here):
```python
window_SQL = "and datetime(visits.visit_time/1000000-11644473600, 'unixepoch', 'localtime') BETWEEN '{window_start}' AND '{window_end}'"
```

And we use the new `history_SQL` like so (this will be in the base `Browser` class, `fetch_history` method):
```python
if window is not None:
    window_part = ""
else:
    window_part = self.window_SQL.format(window_start=start_dt.toisoformat(), window_end=end_dt.toisoformat())
history_SQL = self.history_SQL.format(window_part=window_part)
```

**Comment by Samyak2** [comment]
To clarify, I propose that we add a `window` param to `fetch_history`, so the API becomes:
```python
Firefox().fetch_history(window=[start_dt, end_dt])
```
The actual `history_SQL` property is not modified. A new one is created on the fly for the specific window needed.

**Comment by markelwin** [comment]
@Samyak2 If you do not follow the rest let us discuss until you do before making suggestions about API changes. I will make one more appeal to you to not add the parameter to the `.fetch_history()` function and then accept your decision. Your proposal is wrong. As currently implemented, the `.history_SQL` is an exposed property that allows the user to implement any SQL query for `.fetch_history`. This is the correct behavior and I will continue to use this method in future development of my application, especially as unforeseen use cases may inherent any number of future changes to the SQL Query. Now, on to your specific concern regarding the usability for the novice user. We are in agreement. The novice user should not need to have any knowledge of SQL or that SQL exists in order to be able to access the feature we are adding today. That say, we are debating where to put that line of code that adjusts the SQL to make a windowed query. I am arguing that the query is Browser specific and therefore should be a logical routine on each Browser subclass, and therefore implement the design pattern which adjusts the Browser instance (the `f` object) instead of being hidden within the action method `.fetch_history()`, which it currently is not because `.history_SQL` is an exposed attribute. I am correct. You are arguing that the user should not be required to inherit a change in design pattern logic, and that because users have never needed to manipulate the Browser instance prior to interacting with the design of the `.fetch_history()` action method they should continue to never need to inherit a change in design patter logic, and that is a reasonable case to make, but consider subsequent changes to `.history_SQL` for instance websites which the users spends less than 0.75 seconds on to filter out bad searches or transient backend url requests. How would such a design be implemented? By constructing a change on the `.history_SQL` object directly and then calling `.fetch_history()` without any change of parameters, and in most cases the new parameters added to `.fetch_history()` will need to be ignored even when `.history_SQL` contains windowing for `visit_date`. That sentence is my argument against the proposed change you have asked to implement, and I will accept your decision.

As for implementing the `.history_SQL` logic itself I advise using a slightly different formulation for templating, but for all intents and purposes we are describing the same implementation unless we expect the user to use an SQL constructor.

```python
def SQL_chrome(from_dt=None, to_dt=None):
    # Construct the WHERE clause for visit_time
    strf = "%Y-%m-%d %H:%M:%S"  # '2024-10-06 07:00:00'
    visit_time = "datetime(visits.visit_time/1000000-11644473600, 'unixepoch', 'localtime')"
    if from_dt is None and to_dt is not None:
        insert = f"    {visit_time} <= '{to_dt.strftime(strf)}'\n    AND"
    elif from_dt is not None and to_dt is None:
        insert = f"    {visit_time} >= '{from_dt.strftime(strf)}'\n    AND"
    elif from_dt is not None and to_dt is not None:
        insert = f"    {visit_time} BETWEEN '{from_dt.strftime(strf)}' AND '{to_dt.strftime(strf)}'\n    AND"
    else:
        insert = ""

    return \
f"""
SELECT
    datetime(
        visits.visit_time/1000000-11644473600, 'unixepoch', 'localtime'
    ) as 'visit_time',
    urls.url,
    urls.title
FROM
    visits INNER JOIN urls ON visits.url = urls.id
WHERE
{insert}
    visits.visit_duration > 0
ORDER BY
    visit_time DESC
""".strip()
```

**Comment by Samyak2** [comment]
@markelwin sorry for the late response and thank you for explaining.

---

> As currently implemented, the `.history_SQL` is an exposed property that allows the user to implement any SQL query for `.fetch_history`.

You are right. As it stands now, all attributes of all classes in browser-history are public. This was a mistake. I did not think of it when I first implemented them. I tried to correct it in #194 (among other things, it also makes the `history_SQL` private), but I don't think it will be merged anytime soon (it's an abandoned PR).

Anyway, what I'm trying to say is that `history_SQL` is an implementation detail. It varies across browsers. And if one is constructing the SQL themselves, it will be specific to that browser and does not benefit from the abstractions that browser-history provides.

---

> but consider subsequent changes to `.history_SQL` for instance websites which the users spends less than 0.75 seconds on to filter out bad searches or transient backend url requests. How would such a design be implemented?

Those filters seem specific to a use case, not sure a generic solution to that will look like. But in any case, if it can be done with a separate method (such as `.add_filter("duration", "<", 0.75)`), the same can be done with a param (`.fetch_history(filters=[("duration", "<", 0.75)])`).

---

Another line of reasoning I'm thinking along. Consider the `browser_history.get_history` function. To add support for time filters for that, we would need some params that can be passed to it. And some common method across all browsers that can be used to apply those params. So it would make sense that `get_history` and `browser.fetch_history` have the same API and use the same param names. Instead of a different method in `browser`.

@ObaraEmmanuel thoughts?

**Comment by markelwin** [comment]
Please do not make `. history_SQL` private. Firstly the design choice is not incorrect and provides explicit customization for power users. Secondly to do so is a potentially breaking change for your CustomBrowser extensions. As I said in my uses I will explicitly override the `. history_SQL` as this only requires that the user know SQL and is infinitely expansive. For the implementation of the PR we are describing I will implement your new parameters to the `.fetch_history()` as promised. Not sure what you mean by those filters being specific to a use case: if you are intending on offering near-full SQL functionality in search then simply leave the `.history_SQL` public and users can implement any use case they want. The suggestions you are offering are much much worse to support (consider a change as simple as reversing the order of the `visits.visit_duration > 0` and search time window. If you are concluding that amending the Browser object with the SQL method is not sufficient then I will add the the two parameters to the `.fetch_history()` object. Please finish discussing and I will move my existing code to a PR this weekend. I stand by my original vote and will accept your alternative. Thanks.

**Comment by Samyak2** [comment]
@markelwin you make a good case for allowing custom history SQL. Can you show us how you override the history_SQL? Do you copy-paste the SQL from browser-history's source and then change it to suit your needs?

---

In any case, I believe we can reach an understanding here. Allowing support for custom SQL doesn't need to go against having a simpler API in `fetch_history`. I say we make the feature official and add a `custom_sql` parameter to `fetch_history`. This param will be mutually exclusive with `window` (and any other filtering params we add in the future).

I stand by `.history_SQL` being an internal/private property. I would still consider making it private in the future (of course only if we decide to the implement the `custom_sql` param).

---

How does this sound to you?

**Comment by markelwin** [comment]
My hunch is you do not want a customer SQL parameter to `.fetch_history()`. Notice that in our simple use case the SQL is still dependent on the Browser class. The SQL needed for window search is different for Chrome and Firefox.

**Comment by Samyak2** [comment]
> My hunch is you do not want a customer SQL parameter to `.fetch_history()`. Notice that in our simple use case the SQL is still dependent on the Browser class. The SQL needed for window search is different for Chrome and Firefox.

Yes, the SQL will still be specific to the browser. How is adding an `custom_sql` parameter to `.fetch_history()` any different from assigning to `.history_SQL` in terms of this?

**Comment by ObaraEmmanuel** [comment]
I am with @Samyak2 on this. Anyone that has decided to use a custom SQL query will probably be conversant with which browser they are trying to work with. Adding a custom SQL parameter to `.fetch_history` results in a cleaner API. `.history_SQL` should be deprecated with the introduction of the custom SQL parameter and then be made private over time.

**Comment by markelwin** [comment]
Those are both the wrong decisions. I am sorry, and I cannot agree with you. What I have suggested in a simple inline function call on the browser object to generate the required window sql one layer behind the scenes, in an api which does not create any breaking changes for any custom users, and which is transparent to what the change provides. What you are not suggesting is far worse than the from_ts and to_ts parameters, as you are now requiring users to actually manage the sql creation for each browser in order to use the basic functionality, including the windowing. And furthermore deprecating the history_sql parameter would break legacy code and is needless. Both an awful implementations of a feature. I am sorry. I will not. I have been software engineering for ten years and these are mistakes which would force me to create a new competitor package, or lock in to an old version with my own fork to maintain, neither of which I would want to do. It is not a conversation point we should have reached.

**Comment by Samyak2** [comment]
Okay this is getting a bit heated. I still think there's misunderstanding and that our ideas are not that different. Other than a few details.

---

> What you are not suggesting is far worse than the from_ts and to_ts parameters, as you are now requiring users to actually manage the sql creation for each browser in order to use the basic functionality, including the windowing.

I'm not sure what you're referring to here. I do not see anywhere in our conversation where I suggested making the user manage SQL creation. My position has been the opposite of that. I do not want users to be concerned with the SQL unless they are a power user (who can then use the `custom_sql` parameter if needed).

---

I'm willing to explain my position again. The changes to the API would be as follows:
- In `fetch_history`, add a new **optional** parameter called `window` that takes a tuple of (start_dt, end_dt). I do not mind if this becomes two separate params `window_start`/`window_end` or other names. If this param is provided, we inject the relevant filters in the history_SQL before executing it. The API will look as follows:
    ```python
    f = Firefox()
    outputs = f.fetch_history(window=(start_dt, end_dt))
    ```
- In `browser_history.get_history`, add the same window or similar params.
- Only in `fetch_history`, add a new **optional** parameter called `custom_sql`. If provided, it will override the `history_SQL` and no windowing, etc. will be performed on top of it. The given SQL will be executed as-is. The default behaviour is still the same (since custom_sql is optional).
    ```python
    f = Firefox()
    outputs = f.fetch_history(custom_sql="select ...")
    ```

The existing API will remain as-is. No breaking changes:
```python
f = Firefox()
outputs = f.fetch_history()
```

---

As for deprecating `history_SQL`, let's move that conversation to a separate issue.

**Comment by Samyak2** [comment]
Here's the issue for deprecation of history_SQL: https://github.com/browser-history/browser-history/issues/285

**Comment by ObaraEmmanuel** [comment]
It makes more sense to have separate parameters for `start_dt` or `end_dt`. That way you can request history up to a certain point or starting from a point. Also, I would prefer the simple param names `start` and `end`.

```python
f = Firefox()
outputs = f.fetch_history(start="yyy")

outputs = f.fetch_history(end="yyy")

outputs = f.fetch_history(start="yyy", end="yyy")
```

What date time format would we use here @Samyak2 ?

**Comment by Samyak2** [comment]
@ObaraEmmanuel we can accept datetime objects directly in the API. I don't see any reason to accept a string instead. For the CLI, we can accept the standard [ISO8601 format](https://en.wikipedia.org/wiki/ISO_8601). Python has `datetime.isoformat()` and `datetime.fromisoformat()` for converting between iso strings and datetime objects.

**Comment by Samyak2** [comment]
@markelwin what do you think? Let us know if you're still interested in this. There are some good ideas here which will be very useful improvements to browser-history.
